### PR TITLE
mnstitch - Stitching improvement

### PR DESCRIPTION
**Pull request by biyong**
Root Cause Analysis: Optimization

Fix/Implementation:
	1. Use intrinsic camera matrix from estimator instead of hard coded custom camera matrix
	2. Expose feature detecting tuning parameters through command line
        3. Code clean up

Ticket #: EM-2458

Reviewed By: <reviewers name>

### CODE REVIEW

**Comment by gytibor** [comment]
Have you considered using UMat for the images and masks, so this tool can take advantage of using the GPU for transformations?

File: src/mnstitch_util.cpp [hunk 1]
```diff
@@ -16,313 +16,318 @@ using namespace cv;
...
+
+		// Calculate width and height of the current rectangle.
+		x = cSortedX[minXIdx].x;
+		y = cSortedY[minYIdx].y;
+		int xMax = cSortedX[maxXIdx].x;
+		int yMax = cSortedY[maxYIdx].y;
+		width = xMax - x;
+		height = yMax - y;
+
+#if 0
```
Comment (gytibor) [inline src/mnstitch_util.cpp]: Please consider using a debug or verbose define instead of the #if 0.

File: src/mnstitch_main.cpp [hunk 2]
```diff
@@ -1,285 +1,438 @@
...
+
+	for (int i = 0; i < num_images; i++)
+		resize(source_images[i], images[i], Size(), work_scale, work_scale);
+
+	int64 t = getTickCount();
+	findFeatures(images, features);
+	time.find_features = (getTickCount() - t) / getTickFrequency();
+
+	for (int i = 0; i < num_images; i++)
+		resize(source_images[i], images[i], Size(), seam_scale, seam_scale);
```
Comment (gytibor) [inline src/mnstitch_main.cpp]: The seam scale is always 1 here, so this resize is not needed.
Also this resize overwrite the image with the source images, dropping the effect of the work_scale .

File: src/mnstitch_main.cpp [hunk 3]
```diff
@@ -1,285 +1,438 @@
...
+	//
+	cout << "Finding features ..." << endl;
+	vector<Mat> images(num_images);
+	vector<detail::ImageFeatures> features;
+
+	if (work_megapix > 0)
+		work_scale = min(1.0, sqrt(work_megapix * 1e6 / source_images[0].size().area()));
+
+	for (int i = 0; i < num_images; i++)
+		resize(source_images[i], images[i], Size(), work_scale, work_scale);
```
Comment (gytibor) [inline src/mnstitch_main.cpp]: This resize is only needed if work scale is not 1.

Comment (biyong) [inline src/mnstitch_main.cpp]: That is true that resize is only needed if work scale is not 1. But I don't want to touch the original image, whether or not work scale is 1, I need to new Mat to work on for feature detection.

File: src/mnstitch_main.cpp (hunk 2 above)
Comment (biyong) [inline src/mnstitch_main.cpp]: I missed seam_scale calculation here. I will fix that.

File: src/mnstitch_util.cpp [hunk 4]
```diff
@@ -16,313 +16,318 @@ using namespace cv;
...
+		// Greedy algorithm
+		// Choose the line with the worst deficit and move it toward the center.
+		// Deficit will be (width-count)/width or (height-count)/height so that
+		// the deficits are normalized to the side length.
+		float minXDeficit = (float)(height-minXCount) / (float)height;
+		float maxXDeficit = (float)(height-maxXCount) / (float)height;
+		float minYDeficit = (float)(width-minYCount) / (float)width;
+		float maxYDeficit = (float)(width-maxYCount) / (float)width;
+
+#if 0
```
Comment (gytibor) [inline src/mnstitch_util.cpp]: Same as above.

File: src/mnstitch_util.cpp [hunk 5]
```diff
@@ -16,313 +16,318 @@ using namespace cv;
...
+
+		// Calculate width and height of the current rectangle.
+		x = cSortedX[minXIdx].x;
+		y = cSortedY[minYIdx].y;
+		int xMax = cSortedX[maxXIdx].x;
+		int yMax = cSortedY[maxYIdx].y;
+		width = xMax - x;
+		height = yMax - y;
+
+#if 0
```
Comment (gytibor) [inline src/mnstitch_util.cpp]: Same as above.

File: src/mnstitch_main.cpp [hunk 6]
```diff
@@ -200,6 +200,9 @@ main(int argc, char* argv[])
 	findFeatures(images, features);
 	time.find_features = (getTickCount() - t) / getTickFrequency();
 
+	if (seam_megapix > 0)
+		seam_scale = min(1.0, sqrt(seam_megapix * 1e6 / source_images[0].size().area()));
+
 	for (int i = 0; i < num_images; i++)
 		resize(source_images[i], images[i], Size(), seam_scale, seam_scale);
```
Comment (gytibor) [inline src/mnstitch_main.cpp]: This will still overwrite the source images, discarding the work_scale effects. Is it intentional?

Comment (biyong) [inline src/mnstitch_main.cpp]: work_scale images are consumed in findFeatures(images, features) already.

**Pull request merged**